refactor autograd_hook

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54995 static graph api skeleton
* **#54981 refactor autograd_hook**
* #56730 add a test for grad view with torch amp

put part of codes in autograd_hook into functions, so that they can be used in the static graph training later on.

Differential Revision: [D27439508](https://our.internmc.facebook.com/intern/diff/D27439508/)

zhaojuanmao committed:
refactor autograd_hook

put part of codes in autograd_hook into functions, so that they can be used in the static graph training later on.

Differential Revision: [D27439508](https://our.internmc.facebook.com/intern/diff/D27439508/)

[ghstack-poisoned]

wayi1 reviewed (approved):
Thanks for the refactoring!

rohan-varma commented on torch/lib/c10d/reducer.hpp:
```diff
@@ -422,7 +419,11 @@ class Reducer {
   void search_unused_parameters(
-    const std::vector<torch::autograd::Variable>& outputs);
+      const std::vector<torch::autograd::Variable>& outputs);
+  void set_divide_factor();
```
nit: would be useful to add comments for these functions

rohan-varma reviewed (approved):
Thanks! It's great that we're making reducer code more modular and readable.

zhaojuanmao committed:
Update on "refactor autograd_hook"


put part of codes in autograd_hook into functions, so that they can be used in the static graph training later on.

Differential Revision: [D27439508](https://our.internmc.facebook.com/intern/diff/D27439508/)

[ghstack-poisoned]

zhaojuanmao commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -421,6 +438,14 @@ void Reducer::autograd_hook(VariableIndex index) {
+  // Ignore if we don't expect to be called.
+  // This may be the case if the user wants to accumulate gradients
+  // for number of iterations before reducing them.
+  if (!expect_autograd_hooks_) {
```
I think it is the right place, if sync is not needed, local_used_maps_ does not need to be set as well.

zhaojuanmao committed:
Update on "refactor autograd_hook"


put part of codes in autograd_hook into functions, so that they can be used in the static graph training later on.

Differential Revision: [D27439508](https://our.internmc.facebook.com/intern/diff/D27439508/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "refactor autograd_hook"


put part of codes in autograd_hook into functions, so that they can be used in the static graph training later on.

Differential Revision: [D27439508](https://our.internmc.facebook.com/intern/diff/D27439508/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "refactor autograd_hook"


put part of codes in autograd_hook into functions, so that they can be used in the static graph training later on.

Differential Revision: [D27439508](https://our.internmc.facebook.com/intern/diff/D27439508/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "refactor autograd_hook"


put part of codes in autograd_hook into functions, so that they can be used in the static graph training later on.

Differential Revision: [D27439508](https://our.internmc.facebook.com/intern/diff/D27439508/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "refactor autograd_hook"


put part of codes in autograd_hook into functions, so that they can be used in the static graph training later on.

Differential Revision: [D27439508](https://our.internmc.facebook.com/intern/diff/D27439508/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "refactor autograd_hook"


put part of codes in autograd_hook into functions, so that they can be used in the static graph training later on.

Differential Revision: [D27439508](https://our.internmc.facebook.com/intern/diff/D27439508/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "refactor autograd_hook"


put part of codes in autograd_hook into functions, so that they can be used in the static graph training later on.

Differential Revision: [D27439508](https://our.internmc.facebook.com/intern/diff/D27439508/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "refactor autograd_hook"


put part of codes in autograd_hook into functions, so that they can be used in the static graph training later on.

Differential Revision: [D27439508](https://our.internmc.facebook.com/intern/diff/D27439508/)

[ghstack-poisoned]
